go/vt/vtgate/planbuilder/route.go

```diff
@@ -386,8 +386,25 @@ func (rb *route) JoinCanMerge(pb *primitiveBuilder, rrb *route, ajoin *sqlparser
 		}
 	case engine.SelectReference:
 		return true
-	case engine.SelectNext, engine.SelectDBA:
+	case engine.SelectNext:
 		return false
+	case engine.SelectDBA:
+		if rrb.eroute.Opcode != engine.SelectDBA {
+			return false
+		}
+		if where == nil {
+			return true
+		}
+		hasRuntimeRoutingPredicates := false
+		sqlparser.Walk(func(node sqlparser.SQLNode) (kontinue bool, err error) {
+			col, ok := node.(*sqlparser.ColName)
+			if ok {
+				hasRuntimeRoutingPredicates = hasRuntimeRoutingPredicates ||
+					col.Name.EqualString("table_schema") || col.Name.EqualString("table_name")
```

Comment:
We need to move this to a common place as these are referred at other places as well.